src/mpid/ch4/shm/posix/posix_am_new.h

```diff
@@ -0,0 +1,70 @@
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+
+#ifndef POSIX_AM_H_INCLUDED
+#define POSIX_AM_H_INCLUDED
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_am_send_contig(int rank, int payload_type,
+                                                        int handler_id, int completion_id,
+                                                        int header_size, void *header,
+                                                        int payload_size, void *payload,
+                                                        void *context)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIDI_POSIX_am_header_t msg_hdr;
+    msg_hdr.kind = 0;
```

Comment:
What bugs me about the current code is that this extra header is needed basically for all transports. If we include `handler_id`, `hdr_size`, and `payload_size` in every ch4-level header, then the need for this additional header goes away. Another thing to note is that send completion can be delayed, so `msg_hdr` needs to be on the stack to prevent going out of scope.

Reply:
In fact, `hdr_size` can be implied by the `handler_id`. `payload_size` can be optional depending on the message type.

Reply:
One thing I don't understand is why we are treating context as an opaque object? In the rndv protocol send completion callback (`src/mpid/ch4/am/rndv/rndv.c`) we are using `MPIR_Request`. Is there any case in which context is not a `request`?

Reply:
Not at the moment. I think the intention of making context opaque is that it is not to be interpreted by the transport layer, just handed back to ch4.

Reply:
One possibility is ch4 may define a slimmer context structure if a full request is not needed.

Reply:
> What bugs me about the current code is that this extra header is needed basically for all transports. If we include `handler_id`, `hdr_size`, and `payload_size` in every ch4-level header, then the need for this additional header goes away. Another thing to note is that send completion can be delayed, so `msg_hdr` needs to be on the stack to prevent going out of scope.

I am still in the process of digesting how the current active message works. Note that `msg_hdr` is the transport-layer header and `handler_id`, `hdr_size`, and `payload_size` are there because the transport-layer need that information. The opaque `header` is the protocol-layer header and in principle only contains information that protocol layer needed. For example, the protocol-layer do not need `handler_id` as the correct handler are to-be directly called. It won't need `header_size` either since typically the specific handler knows the specific header structure. It may or may not need `payload_size` depend on the the protocol. For example, if the protocol is chunking the payload to pre-determined chunk-size or have the size pre-communicated in the previous hand-shake, then it don't need payload-size other than to error-check. The worst case, having potentially duplicated information in both headers but separate for different-layer's consumption, IMO, is not detrimental. Leaving both headers separate and opaque to each layer, I think leaves best flexibility and cleaner code.

Reply:
> Note that `msg_hdr` is the transport-layer header and `handler_id`, `hdr_size`, and `payload_size` are there because the transport-layer need that information. 

Why does the transport need that information?

Reply:
Not always, I guess. `handler_id` is needed so the target can call the right handler (`MPIDI_AM_run_handler` is just a interface for the transport layer). In case the transport opt to pack all `msg_hdr`, protocol-layer `header` and `payload` into one packet, then it will need respective sizes in order to resolve them at the target side. I agree that they are not always needed, in which case the `msg_hdr` can omit them. Once the framework is in place, further refactor at the respective transport layer will be in order. At which point, we probably will not include `handler_id`, `hdr_size`, and `payload_size` in every ch4-level header as they are not needed.

Reply:
IMO, ch3 got this right before. If `handler_id` (aka packet type, see https://github.com/pmodels/mpich/blob/master/src/mpid/ch3/include/mpidpkt.h) is required to be the first member of each header struct, then transports can just hand the message up for processing without knowing anything about it. The size of the header will be known by the handler function so it can compute the location of the payload (if any).